### PR TITLE
getThreadHistory: Filter out messages containing error attachments

### LINKED ISSUE #357: Attachments sometimes returned as 'error'
Looks like sometimes when fetching message history, the facebook api will return an error for one of the attachments in a message, which throws an error in the api here: 
https://github.com/Schmavery/facebook-chat-api/blob/master/utils.js#L338 

This is the contents of the stringified attachment: `{"attach_type":"error"}`

I am consistently returned about 17 of these while fetching ~72,000 messages, although a week ago fetching the same messages didn't return any errored attachments.

IMO it would be better for the API to gracefully continue without throwing an error, potentially removing the attachment from the message (or keep it).

### PR DESCRIPTION
**Pull request by ivkos**
Fixes #357

### CODE REVIEW

**Comment by Schmavery** [comment]
Thanks for taking a look at this issue.  I'm not totally sure how this solves any problems, were you able to replicate the problem in #357?  I had thought the exception was being thrown from formatMessage.

In an ideal world would it be nicer if we could just format these errors like we do in listen?

**Comment by ivkos** [comment]
Yes, I do get the same messages with error attachments.

The exception is thrown from _formatAttachment because it doesn't handle
attach_type=error. The PR fixes that by handling this type of attachment
and formatting it, while saving the original attachment(s). As far as I've
tested, the original attachment doesn't contain any useful information
about the error, but I thought it might be a good idea to save them just in
case this changes or behaves differently on other accounts.

The PR also filters out messages that contain error attachments, so they
don't get returned in the result of getThreadHistory. If you want me, I can
leave getThreadHistory as it is, so that it returns everything including
messages with error attachments.



On Dec 16, 2016 04:12, "Avery Morin" <notifications@github.com> wrote:

Thanks for taking a look at this issue. I'm not totally sure how this
solves any problems, were you able to replicate the problem in #357
<https://github.com/Schmavery/facebook-chat-api/issues/357>? I had thought
the exception was being thrown from formatMessage.

In an ideal world would it be nicer if we could just format these errors
like we do in listen?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/Schmavery/facebook-chat-api/pull/360#issuecomment-267499909>,
or mute the thread

**Comment by Schmavery** [comment]
Oh I see now.  Thanks for clarifying!
@bsansouci what do you think, are we good to merge?

**Comment by AlexGustafsson** [comment]
Are there a way of suppressing the warnings thrown?

**Comment by ivkos** [comment]
@AlexGustafsson Yeah, you can set the log level to something higher than "warn", i.e. "error" or "silent". 
```js
api.setOptions({ logLevel: "error" });
```
See [api.setOptions](https://github.com/Schmavery/facebook-chat-api/blob/master/DOCS.md#setOptions) for more info.

@Schmavery @bsansouci Do you think the "warn" level is appropriate for these messages? Should we use something lower, e.g. "info"?

**Comment by Schmavery** [comment]
Warn seems fine to me.
ping @bsansouci merge?

File: utils.js [hunk 1]
```diff
@@ -334,6 +344,15 @@ function _formatAttachment(attachment1, attachment2) {
         height: attachment1.metadata.dimensions.height,
         duration: attachment1.metadata.duration,
       };
+    case "error":
+      return {
+        type: "error",
+
+        // Save error attachments because we're unsure of their format,
+        // and whether there are cases they contain something useful for debugging.
```
Comment (bsansouci) [inline utils.js]: Do errors really not have a specific format? Did you encounter those in the wild? If you would you mind sharing the shape of the object.

**Comment by bsansouci** [comment]
@ivkos Hey, thanks for the PR.
Am I understanding correctly that you're _removing_ messages from the return of `getThreadHistory` if they had an attachment that has an error? That doesn't seem like a good behavior, as it seems like there would be no way for the user of this library to retrieve those specific message. I don't think we should filter the messages in this case.

Supporting the `error` type seems like a good thing though!

**Comment by Schmavery** [comment]
Thanks!

**Pull request merged**